src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs

```diff
@@ -99,14 +99,12 @@ public static partial class JsonSerializer
             }
         }
 
-        private static void CreateDataExtensionProperty(
-            JsonPropertyInfo jsonPropertyInfo,
-            ref ReadStack state)
+        private static void CreateDataExtensionProperty(JsonPropertyInfo jsonPropertyInfo, ref ReadStack state)
         {
             Debug.Assert(jsonPropertyInfo != null);
             Debug.Assert(state.Current.ReturnValue != null);
 
-            IDictionary extensionData = (IDictionary)jsonPropertyInfo.GetValueAsObject(state.Current.ReturnValue);
+            object extensionData = jsonPropertyInfo.GetValueAsObject(state.Current.ReturnValue);
```

Comment:
Why this change? Should we add an Assert for IDictionary?

Reply:
We have a bug here. I'll revert this for now since my reasoning for the change was unrelated to the following statement (it had to do with an abandoned plan to populate dictionaries with reflection):

Our current rules expect a type that implements `IDictionary<string, JsonElement/object>` as the extension data type:

https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L246-L247

If someone used a type that implements one of the allowed interfaces, but doesn't implement `IDictionary`, e.g. a custom type like [`StringToGenericIDictionaryWrapper<JsonElement>`](https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/tests/Serialization/TestClasses.GenericCollections.cs#L867), as the extension data type, we'll get an `InvalidCastException`.

The subsequent assumptions of the the generic parameters of the dictionary will also be incorrect:

https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs#L113-L118.

Maybe all that needs to be done is remove the cast to IDictionary, and assert instead that the type `is` (IDictionary<string, JsonElement/object>), and add a test showing that custom-implemented extension data works fine.

I'll write up an issue for this since we might also decide that custom types are not allowed as extension data, or want to see what behavior JSON.NET allows.